File: crates/yaml-db/src/database/stats.rs
impl super::YamlDatabase {
    /// Get task counts grouped by feature
    pub fn get_feature_stats(&self) -> Vec<crate::GroupStats> {
        use std::collections::HashMap;
        let mut stats: HashMap<String, crate::GroupStats> = HashMap::new();

        for feature in self.features.get_all() {
            stats.insert(
                feature.name.clone(),
                crate::GroupStats {
                    name: feature.name.clone(),
                    display_name: feature.display_name.clone(),
                    total: 0,
                    done: 0,
                    pending: 0,
                    in_progress: 0,
                    blocked: 0,
                    skipped: 0,
                },
            );
        }

        for task in self.tasks.get_all() {
            if let Some(entry) = stats.get_mut(&task.feature) {
                entry.total += 1;
                match task.status {
                    crate::TaskStatus::Done => entry.done += 1,
                    crate::TaskStatus::Pending => entry.pending += 1,
                    crate::TaskStatus::InProgress => entry.in_progress += 1,
                    crate::TaskStatus::Blocked => entry.blocked += 1,
                    crate::TaskStatus::Skipped => entry.skipped += 1,
                }
            }
        }

        let mut result: Vec<_> = stats.into_values().collect();
        result.sort_by(|a, b| a.name.cmp(&b.name));
        result
    }

    /// Get task counts grouped by discipline
    pub fn get_discipline_stats(&self) -> Vec<crate::GroupStats> {
        use std::collections::HashMap;
        let mut stats: HashMap<String, crate::GroupStats> = HashMap::new();

        for discipline in self.disciplines.get_all() {
            stats.insert(
                discipline.name.clone(),
                crate::GroupStats {
                    name: discipline.name.clone(),
                    display_name: discipline.display_name.clone(),
                    total: 0,
                    done: 0,
                    pending: 0,
                    in_progress: 0,
                    blocked: 0,
                    skipped: 0,
                },
            );
        }

        for task in self.tasks.get_all() {
            if let Some(entry) = stats.get_mut(&task.discipline) {
                entry.total += 1;
                match task.status {
                    crate::TaskStatus::Done => entry.done += 1,
                    crate::TaskStatus::Pending => entry.pending += 1,
                    crate::TaskStatus::InProgress => entry.in_progress += 1,
                    crate::TaskStatus::Blocked => entry.blocked += 1,
                    crate::TaskStatus::Skipped => entry.skipped += 1,
                }
            }
        }

        let mut result: Vec<_> = stats.into_values().collect();
        result.sort_by(|a, b| a.name.cmp(&b.name));
        result
    }

    /// Get overall project progress
    pub fn get_project_progress(&self) -> crate::ProjectProgress {
        let tasks = self.tasks.get_all();
        let total = tasks.len() as u32;
        let done = tasks
            .iter()
            .filter(|t| t.status == crate::TaskStatus::Done)
            .count() as u32;
        let percent = if total > 0 {
            (done * 100) / total
        } else {
            0
        };
        crate::ProjectProgress {
            total_tasks: total,
            done_tasks: done,
            progress_percent: percent,
        }
    }

    /// Get sorted unique tags from all tasks
    pub fn get_all_tags(&self) -> Vec<String> {
        let mut tags = std::collections::BTreeSet::new();
        for task in self.tasks.get_all() {
            for tag in &task.tags {
                tags.insert(tag.clone());
            }
        }
        tags.into_iter().collect()
    }
}

File: src/lib/stats.ts
import type { GroupStats, ProjectProgress, SubsystemData, Task } from '@/types/generated'

interface GroupItem {
  name: string
  displayName: string
}

function isTaskRecord(task: unknown): task is Task {
  return Boolean(task && typeof task === 'object')
}

function getTaskTags(task: Task): string[] {
  return Array.isArray(task.tags) ? task.tags : []
}

export function computeSubsystemStats(tasks: Task[], subsystems: SubsystemData[]): Map<string, GroupStats> {
  const statsMap = new Map<string, GroupStats>()

  for (const subsystem of subsystems) {
    statsMap.set(subsystem.name, {
      name: subsystem.name,
      displayName: subsystem.displayName,
      total: 0,
      draft: 0,
      done: 0,
      pending: 0,
      inProgress: 0,
      blocked: 0,
      skipped: 0
    })
  }

  for (const task of tasks) {
    if (!isTaskRecord(task)) continue
    const stats = statsMap.get(task.subsystem)
    if (!stats) continue
    stats.total++
    switch (task.status) {
      case 'draft':
        stats.draft++
        break
      case 'done':
        stats.done++
        break
      case 'pending':
        stats.pending++
        break
      case 'in_progress':
        stats.inProgress++
        break
      case 'blocked':
        stats.blocked++
        break
      case 'skipped':
        stats.skipped++
        break
    }
  }

  return statsMap
}

export function computeDisciplineStats(tasks: Task[], disciplines: GroupItem[]): Map<string, GroupStats> {
  const statsMap = new Map<string, GroupStats>()

  for (const discipline of disciplines) {
    statsMap.set(discipline.name, {
      name: discipline.name,
      displayName: discipline.displayName,
      total: 0,
      draft: 0,
      done: 0,
      pending: 0,
      inProgress: 0,
      blocked: 0,
      skipped: 0
    })
  }

  for (const task of tasks) {
    if (!isTaskRecord(task)) continue
    const stats = statsMap.get(task.discipline)
    if (!stats) continue
    stats.total++
    switch (task.status) {
      case 'draft':
        stats.draft++
        break
      case 'done':
        stats.done++
        break
      case 'pending':
        stats.pending++
        break
      case 'in_progress':
        stats.inProgress++
        break
      case 'blocked':
        stats.blocked++
        break
      case 'skipped':
        stats.skipped++
        break
    }
  }

  return statsMap
}

export function computeProjectProgress(tasks: Task[]): ProjectProgress {
  const actionableTasks = tasks.filter(
    t => isTaskRecord(t) && typeof t.status === 'string' && t.status !== 'draft' && t.status !== 'skipped'
  )
  const totalTasks = actionableTasks.length
  let doneTasks = 0
  for (const task of actionableTasks) {
    if (task.status === 'done') doneTasks++
  }
  const progressPercent = totalTasks > 0 ? Math.floor((doneTasks * 100) / totalTasks) : 0

  return { totalTasks, doneTasks, progressPercent }
}

export function getAllTags(tasks: Task[]): string[] {
  const tagsSet = new Set<string>()

  for (const task of tasks) {
    if (!isTaskRecord(task)) continue
    for (const tag of getTaskTags(task)) {
      tagsSet.add(tag)
    }
  }

  return Array.from(tagsSet).sort()
}
